core/src/main/java/com/linecorp/armeria/server/DefaultRoute.java

```diff
@@ -43,35 +47,58 @@
     private final Set<HttpMethod> methods;
     private final Set<MediaType> consumes;
     private final Set<MediaType> produces;
+    private final List<RoutingPredicate<HttpParameters>> paramPredicates;
+    private final List<Predicate<HttpHeaders>> headerPredicates;
 
     private final String loggerName;
     private final String meterTag;
 
     private final int complexity;
 
+    private final String cacheKey;
+
     DefaultRoute(PathMapping pathMapping, Set<HttpMethod> methods,
-                 Set<MediaType> consumes, Set<MediaType> produces) {
+                 Set<MediaType> consumes, Set<MediaType> produces,
+                 List<RoutingPredicate<HttpParameters>> paramPredicates,
+                 List<RoutingPredicate<HttpHeaders>> headerPredicates) {
         this.pathMapping = requireNonNull(pathMapping, "pathMapping");
         this.methods = Sets.immutableEnumSet(requireNonNull(methods, "methods"));
         this.consumes = ImmutableSet.copyOf(requireNonNull(consumes, "consumes"));
         this.produces = ImmutableSet.copyOf(requireNonNull(produces, "produces"));
+        this.paramPredicates = ImmutableList.copyOf(requireNonNull(paramPredicates, "paramPredicates"));
+        this.headerPredicates = ImmutableList.copyOf(requireNonNull(headerPredicates, "headerPredicates"));
 
-        loggerName = generateLoggerName(pathMapping.loggerName(), methods, consumes, produces);
+        loggerName = generateLoggerName(pathMapping.loggerName(), methods, consumes, produces,
+                                        paramPredicates, headerPredicates);
 
-        meterTag = generateMeterTag(pathMapping.meterTag(), methods, consumes, produces);
+        meterTag = generateMeterTag(pathMapping.meterTag(), methods, consumes, produces,
+                                    paramPredicates, headerPredicates);
 
         int complexity = 0;
         if (!methods.isEmpty()) {
-            complexity++;
+            complexity += 1;
         }
-
         if (!consumes.isEmpty()) {
-            complexity += 2;
+            complexity += 1 << 1;
         }
         if (!produces.isEmpty()) {
-            complexity += 4;
+            complexity += 1 << 2;
+        }
+        if (!paramPredicates.isEmpty()) {
+            complexity += 1 << 3;
+        }
+        if (!headerPredicates.isEmpty()) {
+            complexity += 1 << 4;
         }
         this.complexity = complexity;
+
+        if (paramPredicates.isEmpty() && headerPredicates.isEmpty()) {
```

Comment:
Couldn't we cache when `paramPredicates` is not empty()?
We already add `routingCtx.path()` to the `summary` which is used for `equals` in `DefaultRoutingContext` so if we add `routingCtx.query()` to the `summary`, isn't it cacheable?

Reply:
Yes, you're right. But I'm worried it makes cache hit ratio lower. Also, it'd affect the existing Armeria-based application which doesn't use this feature. 🤔

Reply:
Ah yes, let's leave it as it is then. 😄